File: pset2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <math.h>

float get_average_base_words(int, int);

int main(void)
{
    // Retrieve user's input
    string text = get_string("Text: ");

    int lettersCount = 0;
    int wordsCount = 0;
    int sentencesCount = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        char c = toupper(text[i]);

        if (c >= 'A' && c <= 'Z')
        {
            // Count how many letters are in the text
            lettersCount++;
        }
        else if (c == ' ' || (c == '\'' && text[i + 1] != 's'))
        {
            // Count how many words there are
            wordsCount++;
        }

        if (c == '.' || c == '!' || c == '?')
        {
            // Count how many sentences there are
            sentencesCount++;
        }
    }
    // Count last word
    wordsCount++;
    printf("%i letter(s)\n", lettersCount);
    printf("%i word(s)\n", wordsCount);
    printf("%i sentence(s)\n", sentencesCount);

    float L = get_average_base_words(lettersCount, wordsCount);
    float S = get_average_base_words(sentencesCount, wordsCount);

    float index = (0.0588 * L) - (0.296 * S) - 15.8;

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %.0f\n", index);
    }
}

// Return the average number of letters per 100 words in the text
float get_average_base_words(int val, int words)
{
    return (val * 100) / words;
}

File: pset6/readability/readability.py
from cs50 import get_string
import re
from functools import reduce


def main():
    text = get_string("Text: ")

    # Count letters
    letters = reduce(lambda acc, char: acc + 1 if 'a' <= char.lower() <= 'z' else acc, text, 0)

    # Get a list with all words
    words = text.split(" ")

    # Get a list with all sentences. The second step is necessary to remove the empty string at the end
    sentences = re.split(r'[.!?]', text)
    sentences = sentences[:-1]

    L = avg_per_n_words(letters, len(words))
    S = avg_per_n_words(len(sentences), len(words))

    # Calculate Coleman-Liau index with the values above
    index = (0.0588 * L) - (0.296 * S) - 15.8

    if index < 1:
        print("Before Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(index)}")


# Return the average number of letters/sentences per 100 words in the text
def avg_per_n_words(desired, words):
    return (desired * 100) / words


# Run program
main()